# 입문자를 위한, 파이썬/R 데이터 분석   

## 활용실습(2): (중급분석) Titanic - Machine Learning from Disaster

참고 : https://www.kaggle.com/c/titanic

- `Kaggle 참여방법` 참고용: 

    1) 회원가입 

    2) 경진대회 선정 (종료된 과거 경진대회 참여를 통해 경험쌓는 것부터 시작)
    
    3) 데이터 이해 (EDA(Exploratory Data Analysis); 탐색적 데이터 분석)
    
    4) 평가척도 이해 (Evaluation)
    
    5) 검증방법 및 Feature Engineering
    
    6) 고급기법 적용 (모델 튜닝, 앙상블 등)


`**(중요) 실습전 : 쥬피터노트북 - [Settings] - [Jupiter Lab Light] 로 변경하세요.**`

In [1]:
import pandas as pd

## 데이터 읽어오기

In [2]:
# train.csv 파일을 읽어옵니다. 여기서 PassengerId라는 컬럼을 인덱스(index)로 지정한 뒤, train 변수에 할당합니다.
train = pd.read_csv("data_titanic/train.csv", index_col="PassengerId")

In [3]:
# train 변수에 할당된 데이터의 행렬 사이즈를 출력
train.shape

(891, 11)

In [4]:
# head()로 train 데이터의 상위 5개를 출력
train.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [ ]:
train.tail()

## 데이터 정리 + Pandas 분석

**1. 타이타닉의 train 데이터에서 1) 전체 생존률과 2) 생존자의 총 인원수, 사망자의 총 인원수를 출력해주세요.**

1번(생존률)의 경우 약 38.3%가 나와야 하며, 2번(인원수)의 경우 생존자의 총 인원수는 342명, 사망자의 총 인원수는 549명이 나와야 합니다.

In [5]:
# 타이타닉의 train데이터에서 Survived 컬럼의 평균을 구합니다.
# 현재 0.0 ~ 1.0 사이의 값을 갖을 것이므로, 백분율(%)로 전환하기 위해 100을 곱해줍니다.
print("### TYPE YOUR CODE ###")
train['Survived']

### TYPE YOUR CODE ###


PassengerId
1      0
2      1
3      1
4      1
5      0
      ..
887    0
888    1
889    0
890    1
891    0
Name: Survived, Length: 891, dtype: int64

In [6]:
train['Survived'].mean()

0.3838383838383838

In [7]:
train['Survived'].mean() * 100

38.38383838383838

<hr><span style="color:white">
survived_rate = train["Survived"].mean()
survived_rate = survived_rate * 100
print(f"생존률 = {survived_rate:.1f}%")
</span><hr>

In [ ]:
# pandas의 value_counts를 활용하여 생존자의 총 인원수와 사망자의 총 인원수를 출력합니다.
print("### TYPE YOUR CODE ###")

In [38]:
train['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

<hr><span style="color:white">
train["Survived"].value_counts()
</span><hr>

In [9]:
### 생존률 구하는 또다른 방법 (value_counts 및 len 사용)
print("### TYPE YOUR CODE ###")

### TYPE YOUR CODE ###


In [10]:
train['Survived'].value_counts() / len(train['Survived'])

0    0.616162
1    0.383838
Name: Survived, dtype: float64

<hr><span style="color:white">
train["Survived"].value_counts() / len(train["Survived"])
</span><hr>

**2. Survived 컬럼에 들어가 있는 값을 쉬운 표현으로 바꿔주세요.**

Survived 컬럼에는 0(사망)이라는 값과 1(생존)이라는 값이 있습니다. 이 표현은 직관적이지 않기 때문에, 데이터 분석을 원활하게 하기 위해서 읽기 쉬운 표현을 쓰는 것이 좋습니다.

Survived(Status)라는 새로운 컬럼을 만들어, 이 컬럼에는 0(사망), 1(생존) 정보를, Perish(사망), Survived(생존)이라는 값으로 전환해주세요.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Survived</th>
        <th>Survived(Status)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>0</td>
        <td>Perish</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>Survived</td>
      </tr>
      <tr>
        <th>3</th>
        <td>1</td>
        <td>Survived</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>Survived</td>
      </tr>
      <tr>
        <th>5</th>
        <td>0</td>
        <td>Perish</td>
      </tr>
    </tbody>
</table>

In [ ]:
# Survived 컬럼의 결과값은 0과 1로 표현되는데, Survived 컬럼에 대한 사전 설명이 없으면 이 값이 어떠한 의미를 가지는지 직관적으로 이해하기 어렵습니다.
train["Survived"].head()

In [ ]:
# 먼저 Survived 컬럼이 0인 승객을 색인합니다. 이후 Survived(Status)라는 이름의
# 새로운 컬럼을 만들어 여기에 Perish 라는 값을 넣습니다.

# Survived 컬럼이 1인 승객에 대해서도 수행합니다.
print("### TYPE YOUR CODE ###")

In [11]:
train.loc[train['Survived'] == 0, 'Survived(Status)'] = 'Perish'
train.loc[train['Survived'] == 1, 'Survived(Status)'] = 'Survived'

In [12]:
train

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   
...            ...    ...               ...      ...   ...      ...   
887              0      0            211536  13.0000   NaN        S   
888              0      0            112053  30.0000   B42        S   
889              1      2        W./C. 6607  23.4500   NaN        S   
890              0      0            111369  30.0000  C148        C   
891              0      0            370376   7.7500   NaN        Q   

            Survived(Status)  
PassengerId                   
1                     Perish  
2                   Survived  
3                   Survived  
4                   Survived  
5                     Perish  
...                      ...  
887                   Perish  
888                 Survived  
889                   Perish  
890                 Survived  
891                   Perish  

[891 rows x 12 columns]

In [ ]:
# pandas의 value_counts를 활용하여 생존자의 총 인원수와 사망자의 총 인원수를 출력합니다.
# 생존자의 총 인원수(Survived)은 342명, 사망자의 총 인원수(Perish)는 549명이 나와야 합니다.
print("### TYPE YOUR CODE ###")

In [13]:
train['Survived(Status)'].value_counts()

Perish      549
Survived    342
Name: Survived(Status), dtype: int64

<hr><span style="color:white">
train["Survived(Status)"].value_counts()
</span><hr>

**3. Pclass 컬럼에 들어가 있는 값을 읽기 쉬운 표현으로 바꿔주세요.**

In [ ]:
# pandas의 pivot_table을 활용하여 Pclass별 생존률을 출력합니다.
print("### TYPE YOUR CODE ###")

In [15]:
train.loc[train['Pclass'] == 1, 'Pclass(Status)'] = 'First Class'
train.loc[train['Pclass'] == 2, 'Pclass(Status)'] = 'Business'
train.loc[train['Pclass'] == 3, 'Pclass(Status)'] = 'Economy'

In [16]:
train.groupby('Pclass(Status)')['Survived'].mean()

Pclass(Status)
Business       0.472826
Economy        0.242363
First Class    0.629630
Name: Survived, dtype: float64

<hr><span style="color:white">
pd.pivot_table(data=train, index="Pclass", values="Survived")
</span><hr>

Pclass도 마찬가지로 1, 2, 3이라는 표현은 직관적이지 않기 때문에, 사람이 이해하기 쉬운 표현으로 바꿔주고 싶습니다.

이번에는 **Pclass(Status)**라는 새로운 컬럼을 만들어주세요. 이 컬럼에는 1, 2, 3이 아닌 First Class, Business, Economy 라는 값이 들어가도록 하여, 최종적으로는 다음의 결과가 나오도록 해보세요.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Pclass</th>
        <th>Pclass(Status)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>3</td>
        <td>Economy</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>First Class</td>
      </tr>
      <tr>
        <th>3</th>
        <td>3</td>
        <td>Economy</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>First Class</td>
      </tr>
      <tr>
        <th>5</th>
        <td>3</td>
        <td>Economy</td>
      </tr>
    </tbody>
</table>

In [ ]:
#Pclass(Status)라는 새로운 컬럼에는 1, 2, 3이 아닌 First Class, Business, Economy 라는 값 입력 
print("### TYPE YOUR CODE ###")

In [ ]:
train.loc[train['Pclass'] == 1, 'Pclass(Status)'] = 'First Class'
train.loc[train['Pclass'] == 2, 'Pclass(Status)'] = 'Business'
train.loc[train['Pclass'] == 3, 'Pclass(Status)'] = 'Economy'

<hr><span style="color:white">
train.loc[train["Pclass"] == 1, "Pclass(Status)"] = "First Class"
train.loc[train["Pclass"] == 2, "Pclass(Status)"] = "Business"
train.loc[train["Pclass"] == 3, "Pclass(Status)"] = "Economy"
print(train.shape)
train[["Pclass", "Pclass(Status)"]].head()
</span><hr>

또한 다음의 결과가 나오도록 출력해 보세요.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Survived</th>
      </tr>
      <tr>
        <th>Pclass(Status)</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Business</th>
        <td>0.472826</td>
      </tr>
      <tr>
        <th>Economy</th>
        <td>0.242363</td>
      </tr>
      <tr>
        <th>First Class</th>
        <td>0.629630</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# Pclass(Status) 컬럼에 대한 생존률
print("### TYPE YOUR CODE ###")

In [18]:
pivot_tavle = train.pivot_table(values= 'Survived', index= 'Pclass(Status)')
pivot_tavle

Survived
Pclass(Status)          
Business        0.472826
Economy         0.242363
First Class     0.629630

<hr><span style="color:white">
pd.pivot_table(data=train, index="Pclass(Status)", values="Survived")
</span><hr>

**4. Embarked 컬럼에 들어가 있는 값을 읽기 쉬운 표현으로 바꿔주세요.**

In [ ]:
# Embarked 기준으로 탑승객의 생존률을 분석
print("### TYPE YOUR CODE ###")

In [ ]:
train

In [41]:
pivot_tavle2 = train.pivot_table(values= 'Survived', index= 'Embarked')
pivot_tavle2
pivot_tavle2 * 100

Survived
Embarked           
C         55.357143
Q         38.961039
S         33.695652

<hr><span style="color:white">
pd.pivot_table(data=train, index="Embarked", values="Survived")
</span><hr>

Embarked 컬럼도 마찬가지로 C, S, Q라는 표현은 직관적이지 않습니다. 이해하기 쉽게끔 C는 Cherbourg 라는 표현으로, S는 Southampton 이라는 표현으로, 그리고 Q는 Queenstown 이라는 표현으로 바꾸겠습니다.

앞에서 Survived, Pclass와 마찬가지로, Embarked 컬럼도 **Embarked(Status)**라는 이름의 새로운 컬럼을 만들고, 이 컬럼에는 C, S, Q가 아닌 Cherbourg, Southampton, Queenstown이라는 값이 들어가도록 전환합니다. 최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Embarked</th>
        <th>Embarked(Status)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>S</td>
        <td>Southampton</td>
      </tr>
      <tr>
        <th>2</th>
        <td>C</td>
        <td>Cherbourg</td>
      </tr>
      <tr>
        <th>3</th>
        <td>S</td>
        <td>Southampton</td>
      </tr>
      <tr>
        <th>4</th>
        <td>S</td>
        <td>Southampton</td>
      </tr>
      <tr>
        <th>5</th>
        <td>S</td>
        <td>Southampton</td>
      </tr>
    </tbody>
  </table>

In [44]:
# Embarked(Status)라는 이름의 새로운 컬럼을 만들고, 
# C, S, Q가 아닌 Cherbourg, Southampton, Queenstown이라는 값이 들어가도록 전환
print("### TYPE YOUR CODE ###")

### TYPE YOUR CODE ###


In [45]:
train.loc[train['Embarked'] == 'C', 'Embarked(Status)'] = 'Cherbourg'
train.loc[train['Embarked'] == 'S', 'Embarked(Status)'] = 'Southampton'
train.loc[train['Embarked'] == 'Q', 'Embarked(Status)'] = 'Queenstown'

In [46]:
train[['Embarked', 'Embarked(Status)']].head()

Embarked Embarked(Status)
PassengerId                          
1                  S      Southampton
2                  C        Cherbourg
3                  S      Southampton
4                  S      Southampton
5                  S      Southampton

<hr><span style="color:white">
train.loc[train["Embarked"] == "C", "Embarked(Status)"] = "Cherebourg"
train.loc[train["Embarked"] == "S", "Embarked(Status)"] = "Southampton"
train.loc[train["Embarked"] == "Q", "Embarked(Status)"] = "Queenstown"
train[["Embarked", "Embarked(Status)"]].head()
#train.shape
</span><hr>

그리고 마찬가지로 이번에도 Embarked가 아닌 새롭게 만든 Embarked(Status) 컬럼으로 pivot_table을 만들어주세요. 여기에는 Cherbourg, Southampton, 그리고 Queenstown 에 탑승한 승객의 생존률이 출력됩니다. 최종적으로는 다음의 결과가 나와야 합니다.

 <table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Survived</th>
      </tr>
      <tr>
        <th>Embarked(Status)</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Cherbourg</th>
        <td>0.553571</td>
      </tr>
      <tr>
        <th>Queenstown</th>
        <td>0.389610</td>
      </tr>
      <tr>
        <th>Southampton</th>
        <td>0.336957</td>
      </tr>
    </tbody>
  </table>

In [47]:
# pandas의 pivot_table을 활용하여 Embarked별 생존률을 출력하는데, Embarked 컬럼이 아닌 Embarked(Status) 컬럼을 사용합니다.
print("### TYPE YOUR CODE ###")

### TYPE YOUR CODE ###


In [48]:
pivot_tavle2 = train.pivot_table(values= 'Survived', index= 'Embarked')
pivot_tavle2

Survived
Embarked          
C         0.553571
Q         0.389610
S         0.336957

<hr><span style="color:white">
pd.pivot_table(data=train, index="Embarked(Status)", values="Survived")
</span><hr>

**5. Embarked 컬럼과 Pclass 컬럼을 One Hot Encoding 해주세요**

앞서 작성한 코드를 활용하여 Embarked 컬럼과 Pclass 컬럼을 충분히 분석하였습니다. 
이번에는 이 두 개의 컬럼을 의사결정나무(Decision Tree)와 같은 머신러닝(Machine Learning) 알고리즘에 넣을 수 있도록 [One Hot Encoding](https://hackernoon.com/what-is-one-hot-encoding-why-and-when-do-you-have-to-use-it-e3c6186d008f)을 하도록 합니다. 
먼저 Embarked 컬럼을 바탕으로, 다음의 세 개의 컬럼을 새롭게 만들고 싶습니다.

  1. **Embarked(C)** - Embarked 컬럼 안에 있는 값이 C면 True, 아니면 False가 들어갑니다.
  1. **Embarked(S)** - Embarked 컬럼 안에 있는 값이 S면 True, 아니면 False가 들어갑니다.
  1. **Embarked(Q)** - Embarked 컬럼 안에 있는 값이 Q면 True, 아니면 False가 들어갑니다.

최종적으로는 다음의 결과가 나와야 합니다.

 <table border="1" class="dataframe" style="float: left">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Embarked</th>
        <th>Embarked(Status)</th>
        <th>Embarked(C)</th>
        <th>Embarked(S)</th>
        <th>Embarked(Q)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>2</th>
        <td>C</td>
        <td>Cherbourg</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>3</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>4</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>5</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>6</th>
        <td>Q</td>
        <td>Queenstown</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>7</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>8</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>9</th>
        <td>S</td>
        <td>Southampton</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>10</th>
        <td>C</td>
        <td>Cherbourg</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# Embakred를 기준으로 One Hot Encoding 컬럼을 만들고, 
# 관련된 컬럼들을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [49]:
train['Embarked(C)'] = train['Embarked'] == 'C'
train['Embarked(S)'] = train['Embarked'] == 'S'
train['Embarked(Q)'] = train['Embarked'] == 'Q'

In [50]:
train[['Embarked' , 'Embarked(Status)' ,'Embarked(C)', 'Embarked(S)', 'Embarked(Q)']].head(10)

Embarked Embarked(Status)  Embarked(C)  Embarked(S)  Embarked(Q)
PassengerId                                                                 
1                  S      Southampton        False         True        False
2                  C        Cherbourg         True        False        False
3                  S      Southampton        False         True        False
4                  S      Southampton        False         True        False
5                  S      Southampton        False         True        False
6                  Q       Queenstown        False        False         True
7                  S      Southampton        False         True        False
8                  S      Southampton        False         True        False
9                  S      Southampton        False         True        False
10                 C        Cherbourg         True        False        False

<hr><span style="color:white">
train["Embarked(C)"] = train["Embarked"] == "C"
train["Embarked(S)"] = train["Embarked"] == "S"
train["Embarked(Q)"] = train["Embarked"] == "Q"
#print(train.shape)

train[["Embarked", "Embarked(Status)", "Embarked(C)", "Embarked(S)", "Embarked(Q)"]].head(10)
</span><hr>

Pclass 컬럼도 비슷하게 다음의 세 개의 컬럼이 만들어져야 합니다.

  1. **Pclass(F)** - Pclass 컬럼 안에 있는 값이 1이면 True, 아니면 False가 들어갑니다.
  1. **Pclass(B)** - Pclass 컬럼 안에 있는 값이 2면 True, 아니면 False가 들어갑니다.
  1. **Pclass(E)** - Pclass 컬럼 안에 있는 값이 3이면 True, 아니면 False가 들어갑니다.
  
최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Pclass</th>
        <th>Pclass(Status)</th>
        <th>Pclass(F)</th>
        <th>Pclass(B)</th>
        <th>Pclass(E)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>First Class</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>3</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>First Class</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>5</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>6</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>7</th>
        <td>1</td>
        <td>First Class</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>8</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>9</th>
        <td>3</td>
        <td>Economy</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>10</th>
        <td>2</td>
        <td>Business</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# Pclass를 기준으로 One Hot Encoding 컬럼을 만들고, 
# 관련된 컬럼들을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [51]:
train['Pclass(F)'] = train['Pclass'] == 1
train['Pclass(B)'] = train['Pclass'] == 2
train['Pclass(E)'] = train['Pclass'] == 3

In [52]:
train[['Pclass' , 'Pclass(Status)' ,'Pclass(F)', 'Pclass(B)', 'Pclass(E)']].head(10)

Pclass Pclass(Status)  Pclass(F)  Pclass(B)  Pclass(E)
PassengerId                                                        
1                 3        Economy      False      False       True
2                 1    First Class       True      False      False
3                 3        Economy      False      False       True
4                 1    First Class       True      False      False
5                 3        Economy      False      False       True
6                 3        Economy      False      False       True
7                 1    First Class       True      False      False
8                 3        Economy      False      False       True
9                 3        Economy      False      False       True
10                2       Business      False       True      False

<hr><span style="color:white">
train["Pclass(F)"] = train["Pclass"] == 1
train["Pclass(B)"] = train["Pclass"] == 2
train["Pclass(E)"] = train["Pclass"] == 3
#print(train.shape)

train[["Pclass", "Pclass(Status)", "Pclass(F)", "Pclass(B)", "Pclass(E)"]].head(10)
</span><hr>

## 나이(Age) 컬럼 분석

**6. 나이(Age) 컬럼에서 다음의 정보를 출력해주세요.**

  * 평균(mean)
  * 표준편차(standard deviation, 이하 std) 
  * 가장 나이가 많은 사람.
  * 가장 나이가 적은 사람.
  
가령 평균은 약 29.7세, 표준편차(std)는 약 14.52가 나와야 합니다.

In [ ]:
# 나이(Age) 컬럼에서 mean 함수를 통해 평균 나이를 구합니다.
# 평균 나이가 약 29.7세라는 것을 알 수 있습니다.
print("### TYPE YOUR CODE ###")

In [27]:
train['Age'].mean()

29.69911764705882

<hr><span style="color:white">
train["Age"].mean()
</span><hr>

In [ ]:
# 나이(Age) 컬럼에서 std 함수를 통해 나이의 표준 편차를 구합니다.
# 나이의 std 값이 약 14.52세라는 것을 알 수 있습니다.
print("### TYPE YOUR CODE ###")

In [28]:
train['Age'].std()

14.526497332334044

<hr><span style="color:white">
train["Age"].std()
</span><hr>

In [ ]:
# 나이(Age) 컬럼에서 min 함수를 통해 나이의 최소치를 구합니다.
# Titanic 에 탑승한 가장 어린 승객은 약 0.42세(생후 4개월 정도)라는 것을 알 수 있습니다.
print("### TYPE YOUR CODE ###")

In [29]:
train['Age'].min()

0.42

<hr><span style="color:white">
train["Age"].min()
</span><hr>

In [ ]:
# 나이(Age) 컬럼에서 max 함수를 통해 나이의 최대치 구합니다.
# 타이타닉호에 탑승한 가장 나이가 많은 승객은 80세라는 것을 알 수 있습니다.
print("### TYPE YOUR CODE ###")

In [30]:
train['Age'].max()

80.0

<hr><span style="color:white">
train["Age"].max()
</span><hr>

또는, pandas의 [describe](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.describe.html)를 활용하면 한 줄의 코드로 평균, 분산, 최소치, 최대치를 볼 수 있습니다.

In [ ]:
# 나이(Age) 컬럼에 대해 describe 함수를 사용합니다.
# 이 함수는 특정 컬럼의 평균, 표준편차(std), 최대치, 최소치와 같은 기초적인 통계치를 보여줍니다.
print("### TYPE YOUR CODE ###")

In [31]:
train['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64


<hr><span style="color:white">
train["Age"].describe()
</span><hr>

**7. 객실 등급별 나이(Age) 컬럼의 평균을 보여주세요.**

이번에는 전체 평균이 아닌 객실 등급(Pclass)별 평균을 보고 싶습니다.

가령 전체 승객의 평균 나이는 약 29.7세이지만, 1등급 승객의 평균 나이는 약 38.2세가 나와야 합니다. 비슷한 방식으로 2등급과 3등급 승객의 평균 나이를 알 수 있다면 좋겠습니다.

In [ ]:
# 1등급 객실 승객의 평균 나이를 구합니다.
print("### TYPE YOUR CODE ###")

In [53]:
train.loc[train['Pclass'] == 1, 'Age'].mean()

38.233440860215055

<hr><span style="color:white">
train[train["Pclass"]==1]["Age"].mean()
</span><hr>

In [ ]:
# 2등급 객실 승객의 평균 나이를 구합니다. 
print("### TYPE YOUR CODE ###")

In [54]:
train.loc[train['Pclass'] == 2, 'Age'].mean()

29.87763005780347

<hr><span style="color:white">
train[train["Pclass"]==2]["Age"].mean()
</span><hr>

In [ ]:
# 3등급 객실 승객의 평균 나이를 구합니다.
print("### TYPE YOUR CODE ###")

In [55]:
train.loc[train['Pclass'] == 3, 'Age'].mean()

25.14061971830986

<hr><span style="color:white">
train[train["Pclass"]==3]["Age"].mean()
</span><hr>

또는, pandas의 [groupby](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html)를 활용하면 한 줄의 코드로 1, 2, 3등급 객실 승객의 평균 나이를 가져올 수 있습니다.

In [ ]:
# Pclass 기준으로 그룹화하여, 나이(Age) 컬럼의 평균(mean)을 구하면 1, 2, 3등급 객실마다의 평균 나이를 구할 수 있습니다.
print("### TYPE YOUR CODE ###")

In [56]:
train.groupby('Pclass')['Age'].mean()

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

<hr><span style="color:white">
train.groupby("Pclass")["Age"].mean()
pd.pivot_table(data=train, index="Pclass", values="Age")
</span><hr>

**8. 나이를 일정 구역(범위)으로 나눠서, 구역(범위)마다의 생존률을 보여주세요.**

이번에는 나이(Age)별 생존률을 확인하고 싶습니다. 다만 나이 컬럼은 숫자이기 때문에, 그대로 쓰지 않고 일정 구역으로 나눈 뒤 생존률의 통계를 내는 것이 보기 편할 것입니다. 나이 컬럼을 다음의 세 구역으로 나눠봅니다.

  1. Young : 나이가 15세 미만인 승객.
  2. Medium : 나이가 15세 이상이고 30세 미만인 승객.
  3. Old : 나이가 30세 이상인 승객.

최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Age</th>
        <th>AgeType</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>22.0</td>
        <td>Medium</td>
      </tr>
      <tr>
        <th>2</th>
        <td>38.0</td>
        <td>Old</td>
      </tr>
      <tr>
        <th>3</th>
        <td>26.0</td>
        <td>Medium</td>
      </tr>
      <tr>
        <th>4</th>
        <td>35.0</td>
        <td>Old</td>
      </tr>
      <tr>
        <th>5</th>
        <td>35.0</td>
        <td>Old</td>
      </tr>
      <tr>
        <th>6</th>
        <td>NaN</td>
        <td>NaN</td>
      </tr>
      <tr>
        <th>7</th>
        <td>54.0</td>
        <td>Old</td>
      </tr>
      <tr>
        <th>8</th>
        <td>2.0</td>
        <td>Young</td>
      </tr>
      <tr>
        <th>9</th>
        <td>27.0</td>
        <td>Medium</td>
      </tr>
      <tr>
        <th>10</th>
        <td>14.0</td>
        <td>Young</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 나이(Age) 컬럼과 AgeType 컬럼을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [57]:
train.loc[train['Age'] < 15, 'AgeType'] = 'Young'
train.loc[(train['Age'] >= 15) & (train['Age'] < 30), 'AgeType'] = 'Medium'
train.loc[train['Age'] >= 30, 'AgeType'] = 'Old'

<hr><span style="color:white">
train.loc[train["Age"] < 15, "AgeType"] = "Young"
train.loc[(train["Age"] >= 15) & (train["Age"] < 30), "AgeType"] = "Medium"
train.loc[train["Age"] >= 30, "AgeType"] = "Old"

train[["Age", "AgeType"]].head(10)
</span><hr>

이번엔, 위 조건에서 AgeType별 해당하는 승객의 평균 생존률을 구해봅니다. (예: 구역(범위)가 Young 인 경우, 약 57.7%가 나와야 합니다.)

In [ ]:
# AgeType별 해당하는 승객의 평균 생존률
print("### TYPE YOUR CODE ###")

In [58]:
train.groupby('AgeType')['Survived'].mean() * 100

AgeType
Medium    36.274510
Old       40.606061
Young     57.692308
Name: Survived, dtype: float64

<hr><span style="color:white">
train.groupby("AgeType")["Survived"].mean()
#또다른 방법
pd.pivot_table(data=train, index="AgeType", values="Survived")
</span><hr>

**9. 나이가 비어있는 승객과 비어있지 않은 승객의 생존률 차이를 보여주세요.**

이번에는 다른 방식으로 생존률의 차이를 살펴봅니다. Titanic 데이터의 나이(Age) 컬럼을 살펴보면 Age 가 비어있는 데이터가 있는데, 이를 NaN(Not a Number의 약자)으로 표현합니다.

Titanic 데이터에서 Age 컬럼이 비어있는 승객과 비어있지 않은 승객의 생존률을 각각 찾아서 출력해보세요.

In [ ]:
# isnull 함수와, notnull 함수를 활용하여, AgeBlank 컬럼에 "Blank" 와 "Not Blank" 값을 넣어봅니다.
# 나이(Age) 컬럼과 AgeBlank 컬럼을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [59]:
import numpy as np

In [60]:
train['AgeBlank'] = np.where(train['Age'].isnull(), 'Blank', 'Not Blank')

<hr><span style="color:white">
train.loc[train["AgeType"].isnull(), "AgeBlank"] = "Blank"
train.loc[train["AgeType"].notnull(), "AgeBlank"] = "Not Blank"
#train.loc[~train["AgeType"].isnull(), "AgeBlank"] = "Not Blank"

train[["Age", "AgeBlank"]].head(10)
</span><hr>

In [ ]:
# Age 컬럼이 비어있는 승객과 비어있지 않은 승객의 생존률
print("### TYPE YOUR CODE ###")

In [63]:
train.groupby('AgeBlank')['Survived'].mean()

AgeBlank
Blank        0.293785
Not Blank    0.406162
Name: Survived, dtype: float64

<hr><span style="color:white">
train.groupby("AgeBlank")["Survived"].mean()
</span><hr>

**10. Pclass별 나이(Age)의 평균을 구한 뒤 빈 값에 채워주세요.**

이번에는 나이(Age) 컬럼의 빈 값을 채우려고 합니다. 
일반적으로 가장 많이 하는 방식은 평균(mean)값을 구한 뒤, 이를 빈 값에 채워넣는 것인데, 이번에는 전체 평균 나이가 아닌, 각 객실 등급(Pclass)별 평균 연령을 구해서 해당 객실 등급에 빈 Age 값을 채우도록 하겠습니다.

  1. 객실 등급(Pclass)이 1등급인 승객의 평균 나이를 구해서, 해당 승객 중 나이(Age)컬럼값이 비어있는 승객을 찾아 빈 나이 값을 채워줍니다.
  2. 객실 등급(Pclass)이 2등급인 승객의 평균 나이를 구해서, 해당 승객 중 나이(Age)컬럼값이 비어있는 승객을 찾아 빈 나이 값을 채워줍니다.
  3. 객실 등급(Pclass)이 3등급인 승객의 평균 나이를 구해서, 해당 승객 중 나이(Age)컬럼값이 비어있는 승객을 찾아 빈 나이 값을 채워줍니다.
  
위와 같은 방식을 사용하면, 단순히 전체 평균을 사용하는 것 보다 조금 더 원래 값에 근접하게 평균을 채워줄 수 있을 것 같습니다. 최종적으로는 다음의 결과가 나와야 합니다.

 <table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Pclass</th>
        <th>Age</th>
        <th>Age(fill)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>3</td>
        <td>22.0</td>
        <td>22.00000</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>38.0</td>
        <td>38.00000</td>
      </tr>
      <tr>
        <th>3</th>
        <td>3</td>
        <td>26.0</td>
        <td>26.00000</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>35.0</td>
        <td>35.00000</td>
      </tr>
      <tr>
        <th>5</th>
        <td>3</td>
        <td>35.0</td>
        <td>35.00000</td>
      </tr>
      <tr>
        <th>6</th>
        <td>3</td>
        <td>NaN</td>
        <td>25.14062</td>
      </tr>
      <tr>
        <th>7</th>
        <td>1</td>
        <td>54.0</td>
        <td>54.00000</td>
      </tr>
      <tr>
        <th>8</th>
        <td>3</td>
        <td>2.0</td>
        <td>2.00000</td>
      </tr>
      <tr>
        <th>9</th>
        <td>3</td>
        <td>27.0</td>
        <td>27.00000</td>
      </tr>
      <tr>
        <th>10</th>
        <td>2</td>
        <td>14.0</td>
        <td>14.00000</td>
      </tr>
      <tr>
        <th>11</th>
        <td>3</td>
        <td>4.0</td>
        <td>4.00000</td>
      </tr>
      <tr>
        <th>12</th>
        <td>1</td>
        <td>58.0</td>
        <td>58.00000</td>
      </tr>
      <tr>
        <th>13</th>
        <td>3</td>
        <td>20.0</td>
        <td>20.00000</td>
      </tr>
      <tr>
        <th>14</th>
        <td>3</td>
        <td>39.0</td>
        <td>39.00000</td>
      </tr>
      <tr>
        <th>15</th>
        <td>3</td>
        <td>14.0</td>
        <td>14.00000</td>
      </tr>
      <tr>
        <th>16</th>
        <td>2</td>
        <td>55.0</td>
        <td>55.00000</td>
      </tr>
      <tr>
        <th>17</th>
        <td>3</td>
        <td>2.0</td>
        <td>2.00000</td>
      </tr>
      <tr>
        <th>18</th>
        <td>2</td>
        <td>NaN</td>
        <td>29.87763</td>
      </tr>
      <tr>
        <th>19</th>
        <td>3</td>
        <td>31.0</td>
        <td>31.00000</td>
      </tr>
      <tr>
        <th>20</th>
        <td>3</td>
        <td>NaN</td>
        <td>25.14062</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# Pclass 기준으로 그룹화한 뒤, 나이(Age) 컬럼의 평균을 구합니다.
print("### TYPE YOUR CODE ###")

In [64]:
Age_mean = train.groupby('Pclass')['Age'].mean()
Age_mean

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

<hr><span style="color:white">
train.groupby("Pclass")["Age"].mean()
</span><hr>

In [ ]:
# Age 컬럼에 바로 값을 채워주는 것도 좋지만, 가능한 원본은 유지하고, 추가 컬럼에다가 작업하는 것을 추천합니다.
# 여기서는 Age(fill) 이라는 새로운 컬럼을 만든 뒤 원본 데이터를 채워줍니다.
print("### TYPE YOUR CODE ###")

In [65]:
train['Age(fill)'] = train['Age'].fillna(Age_mean)

In [66]:
Age_mean = train.groupby('Pclass')['Age'].transform('mean')
train['Age(fill)'] = train['Age'].fillna(Age_mean)

<hr><span style="color:white">
train.loc[(train['Age'].isnull()) & (train["Pclass"]==1), 'Age(fill)'] = train[train["Pclass"]==1]["Age"].mean()
train.loc[(train['Age'].isnull()) & (train["Pclass"]==2), 'Age(fill)'] = train[train["Pclass"]==2]["Age"].mean()
train.loc[(train['Age'].isnull()) & (train["Pclass"]==3), 'Age(fill)'] = train[train["Pclass"]==3]["Age"].mean()
#print(train.shape)
train[["Pclass", "Age", "Age(fill)"]].head(10)
</span><hr>

In [ ]:
# 작업이 정상적으로 되었는지 확인하기 위하여, 나이(Age) 컬럼값이 비어있는 승객만 가져온 뒤,
# 이 승객의 객실 등급(Pclass), 나이(Age), 그리고 Age(fill) 컬럼을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [67]:
(train[['Pclass', 'Age', 'Age(fill)']]).head(20)

Pclass   Age  Age(fill)
PassengerId                         
1                 3  22.0   22.00000
2                 1  38.0   38.00000
3                 3  26.0   26.00000
4                 1  35.0   35.00000
5                 3  35.0   35.00000
6                 3   NaN   25.14062
7                 1  54.0   54.00000
8                 3   2.0    2.00000
9                 3  27.0   27.00000
10                2  14.0   14.00000
11                3   4.0    4.00000
12                1  58.0   58.00000
13                3  20.0   20.00000
14                3  39.0   39.00000
15                3  14.0   14.00000
16                2  55.0   55.00000
17                3   2.0    2.00000
18                2   NaN   29.87763
19                3  31.0   31.00000
20                3   NaN   25.14062

In [ ]:
### 또다른 방법 : groupby 및 transform 함수 사용
print("### TYPE YOUR CODE ###")

In [99]:
Age_mean = train.groupby('Pclass')['Age'].transform('mean')
train['Age(fill)'] = train['Age'].fillna(Age_mean)

<hr><span style="color:white">
train['Age(fill)'] = train['Age']
train['Age(fill)'].fillna(train.groupby('Pclass')['Age(fill)'].transform('mean'), inplace=True)
train.loc[train["Age"].isnull(), ["Pclass","Age", "Age(fill)"]].head(10)
</span><hr>

## SibSp, Parch 컬럼 분석

**11. 타이타닉호에 동승한 형제, 자매, 배우자(SibSp)도 없고, 부모와 자식(Parch)도 없는 사람을 구해주세요.**

해당 사용자를 싱글(Single)이라고 가정하겠습니다. 최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>SibSp</th>
        <th>Parch</th>
        <th>Single</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>1</td>
        <td>0</td>
        <td>False</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>0</td>
        <td>False</td>
      </tr>
      <tr>
        <th>3</th>
        <td>0</td>
        <td>0</td>
        <td>True</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>0</td>
        <td>False</td>
      </tr>
      <tr>
        <th>5</th>
        <td>0</td>
        <td>0</td>
        <td>True</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# SibSp가 0이고 Parch가 0이면 True, 아니면 False인 리스트를 생성합니다.
# 이 리스트를 Single이라는 이름의 새로운 컬럼에 집어넣습니다.
# SibSp, Parch, 그리고 Single을 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [68]:
train['Single'] = (train['SibSp'] == 0) & (train['Parch'] == 0)

In [69]:
(train[['SibSp', 'Parch', 'Single']]).head(5)

SibSp  Parch  Single
PassengerId                      
1                1      0   False
2                1      0   False
3                0      0    True
4                1      0   False
5                0      0    True

또한 싱글(Single)인 사람과 그렇지 않은 사람간의 생존률의 차이도 알고 싶습니다. 최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Survived</th>
      </tr>
      <tr>
        <th>Single</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>False</th>
        <td>0.505650</td>
      </tr>
      <tr>
        <th>True</th>
        <td>0.303538</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# pandas의 pivot_table을 활용하여 Single 여부에 따른 생존률을 출력합니다.
# Single 컬럼의 값이 True일 경우와, False일 경우의 생존률을 비교할 수 있습니다.
print("### TYPE YOUR CODE ###")

In [70]:
pivot_tavle3 = train.pivot_table(values= 'Survived', index= 'Single')
pivot_tavle3


Survived
Single          
False   0.505650
True    0.303538

<hr><span style="color:white">
pd.pivot_table(data=train, index="Single", values="Survived")
</span><hr>

**12. SibSp 컬럼과  Parch 컬럼을 활용하여 가족 수(FamilySize)라는 새로운 컬럼을 만들어주세요.**

형제, 자매, 배우자(SibSp) 컬럼과 부모 자식(Parch) 컬럼은 얼핏 달라 보이지만 실은 가족 관계를 나타내는 것이라고 볼 수 있습니다. 그러므로 두 컬럼을 하나로 합쳐서 **가족 수(FamilySize)**라는 새로운 컬럼을 만들면 승객의 가족관계를 더 편리하게 분석할 수 있을 것입니다.

형제, 자매, 배우자(SibSp) 컬럼과 부모 자식(Parch) 컬럼을 더해서 가족 수(FamilySize) 컬럼을 만들어주세요. 단 가족 수를 계산할때는 언제나 나 자신을 포함해서 계산하는데, 나 자신은 SibSp 컬럼에도 Parch 컬럼에도 들어가있지 않습니다. 그러므로 가족 수(FamilySize) 컬럼은 언제나 SibSp 컬럼과 Parch 컬럼을 합한 값에서 1을 추가로 더해야 합니다.

그러므로 최종적으로 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>SibSp</th>
        <th>Parch</th>
        <th>FamilySize</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>1</td>
        <td>0</td>
        <td>2</td>
      </tr>
      <tr>
        <th>2</th>
        <td>1</td>
        <td>0</td>
        <td>2</td>
      </tr>
      <tr>
        <th>3</th>
        <td>0</td>
        <td>0</td>
        <td>1</td>
      </tr>
      <tr>
        <th>4</th>
        <td>1</td>
        <td>0</td>
        <td>2</td>
      </tr>
      <tr>
        <th>5</th>
        <td>0</td>
        <td>0</td>
        <td>1</td>
      </tr>
      <tr>
        <th>6</th>
        <td>0</td>
        <td>0</td>
        <td>1</td>
      </tr>
      <tr>
        <th>7</th>
        <td>0</td>
        <td>0</td>
        <td>1</td>
      </tr>
      <tr>
        <th>8</th>
        <td>3</td>
        <td>1</td>
        <td>5</td>
      </tr>
      <tr>
        <th>9</th>
        <td>0</td>
        <td>2</td>
        <td>3</td>
      </tr>
      <tr>
        <th>10</th>
        <td>1</td>
        <td>0</td>
        <td>2</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 형제, 자매, 배우자(SibSp) 컬럼과 부모 자식(Parch) 컬럼을 더해서 가족 수(FamilySize) 컬럼을 만듭니다.
# 또한 가족 수에 나 자신을 포함하기 위해서 언제나 +1을 해줍니다.
# SibSp, Parch, FamilySize를 출력하여 비교합니다.
print("### TYPE YOUR CODE ###")

In [71]:
train['FamilySize'] = (train['SibSp']) + (train['Parch']) +1

In [72]:
(train[['SibSp', 'Parch', 'FamilySize']]).head(10)

SibSp  Parch  FamilySize
PassengerId                          
1                1      0           2
2                1      0           2
3                0      0           1
4                1      0           2
5                0      0           1
6                0      0           1
7                0      0           1
8                3      1           5
9                0      2           3
10               1      0           2

또한 가족 수(FamilySize) 컬럼을 구한 뒤, 가족 수 별 생존률의 차이도 알고 싶습니다. 가족 수(ex: 1명 ~ 11명) 마다의 생존률을 구해서 출력해주세요. 

In [ ]:
# pandas의 pivot_table을 활용하여 FamilySize에 따른 생존률을 출력합니다.
# 가족 수가 1명부터 11명까지 각각의 생존률을 비교할 수 있습니다.
print("### TYPE YOUR CODE ###")

In [73]:
pivot_tavle4 = train.pivot_table(values= 'Survived', index= 'FamilySize')
pivot_tavle4

Survived
FamilySize          
1           0.303538
2           0.552795
3           0.578431
4           0.724138
5           0.200000
6           0.136364
7           0.333333
8           0.000000
11          0.000000

In [74]:
pivot_tavle4 * 100

Survived
FamilySize           
1           30.353818
2           55.279503
3           57.843137
4           72.413793
5           20.000000
6           13.636364
7           33.333333
8            0.000000
11           0.000000

<hr><span style="color:white">
pd.pivot_table(data=train, index="FamilySize", values="Survived")
</span><hr>

**13. 가족 수(FamilySize) 컬럼을 기초로 구역(범위)을 나눠주세요.**

가족 수(FamilySize) 컬럼을 기준으로 pivot_table로 분석을 해본 결과, 경우의 수가 너무 잘게 쪼개져(가족 수가 1명 ~ 11명 범위) 있기 때문에 의미있는 구역(범위) 으로 재조정 하고자 합니다.

가족 수(FamilySize) 컬럼을 3개 구역(범위)으로 다음과 같이 나눠봅니다.

  * **싱글(Single)** - 동승한 가족이 아무도 없고, 나 혼자 탑승한 경우.
  * **핵가족(Nuclear)** - 동승한 가족이 나 자신을 포함해 2명 이상 5명 미만인 경우.
  * **대가족(Big)** - 동승한 가족이 나 자신을 포함 5명 이상인 경우.
  
위의 기준을 바탕으로, 가족 형태(FamilyType)라는 새로운 컬럼을 만들어 주세요. 이 컬럼에는 앞서 설명한 Single, Nuclear, 그리고 Big이 들어갑니다. 최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>FamilySize</th>
        <th>FamilyType</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>2</td>
        <td>Nuclear</td>
      </tr>
      <tr>
        <th>2</th>
        <td>2</td>
        <td>Nuclear</td>
      </tr>
      <tr>
        <th>3</th>
        <td>1</td>
        <td>Single</td>
      </tr>
      <tr>
        <th>4</th>
        <td>2</td>
        <td>Nuclear</td>
      </tr>
      <tr>
        <th>5</th>
        <td>1</td>
        <td>Single</td>
      </tr>
      <tr>
        <th>6</th>
        <td>1</td>
        <td>Single</td>
      </tr>
      <tr>
        <th>7</th>
        <td>1</td>
        <td>Single</td>
      </tr>
      <tr>
        <th>8</th>
        <td>5</td>
        <td>Big</td>
      </tr>
      <tr>
        <th>9</th>
        <td>3</td>
        <td>Nuclear</td>
      </tr>
      <tr>
        <th>10</th>
        <td>2</td>
        <td>Nuclear</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 가족 수(FamilSize)가 1인 승객을 가져와서, FamilyType 컬럼에 Single 이라는 값을 넣어줍니다.
# 가족 수(FamilSize)가 2 이상 5 미만인 승객을 가져와서, FamilyType 컬럼에 Nuclear(핵가족) 이라는 값을 넣어줍니다.
# 가족 수(FamilSize)가 5 이상인 승객을 가져와서, FamilyType 컬럼에 Big(대가족) 이라는 값을 넣어줍니다.
# train 데이터의 상위 10개를 띄우되, FamilySize와 FamilyType 컬럼만 출력합니다.
print("### TYPE YOUR CODE ###")

In [75]:
train.loc[train['FamilySize'] == 1, 'FamilyType'] = 'Single'
train.loc[(train['FamilySize'] >= 2) & (train['FamilySize'] < 5), 'FamilyType'] = 'Nuclear'
train.loc[train['FamilySize'] >= 5, 'FamilyType'] = 'Big'

In [76]:
(train[['FamilySize', 'FamilyType']]).head(10)

FamilySize FamilyType
PassengerId                       
1                     2    Nuclear
2                     2    Nuclear
3                     1     Single
4                     2    Nuclear
5                     1     Single
6                     1     Single
7                     1     Single
8                     5        Big
9                     3    Nuclear
10                    2    Nuclear

또한 가족 수(FamilySize)와 마찬가지로 가족 형태(FamilyType) 별 생존률의 차이도 구해주세요. 최종적으로 다음의 결과가 나와야 합니다.

In [ ]:
# pivot_table을 통해 가족 형태(FamilyType)의 변화에 따른 생존률을 출력합니다.
print("### TYPE YOUR CODE ###")

In [77]:
pivot_tavle5 = train.pivot_table(values= 'Survived', index= 'FamilyType')
pivot_tavle5

Survived
FamilyType          
Big         0.161290
Nuclear     0.578767
Single      0.303538

In [78]:
pivot_tavle5 * 100

Survived
FamilyType           
Big         16.129032
Nuclear     57.876712
Single      30.353818

<hr><span style="color:white">
pd.pivot_table(data=train, index="FamilyType", values="Survived")
</span><hr>

**14. Single, Nuclear, Big 을 각각 One Hot Encoding 해주세요**

앞서 작성한 코드를 분석한 결과, 가족 형태(FamilyType) 컬럼의 값이 생존자와 사망자를 판가름하는데 굉장히 중요한 영향을 차지하는 것 같습니다. 그러므로 이 컬럼값을 이번에는 이 두 개의 컬럼을 의사결정나무(Decision Tree)와 같은 머신러닝(Machine Learning) 알고리즘에 넣을 수 있도록 [One Hot Encoding](https://hackernoon.com/what-is-one-hot-encoding-why-and-when-do-you-have-to-use-it-e3c6186d008f) 처리를 하려고 합니다.

먼저 가족 수(FamilySize)를 바탕으로, 다음의 세 컬럼을 추가로 만들고 싶습니다.

  1. **FamilyType(Single)** - FamilySize가 1이면 True, 아니면 False가 들어갑니다.
  1. **FamilyType(Nuclear)** - FamilySize가 2 이상 5 미만이면 True, 아니면 False가 들어갑니다.
  1. **FamilyType(Big)** - FamilySize가 5 이상이면 True, 아니면 False가 나와야 합니다.

최종적으로는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>FamilySize</th>
        <th>FamilyType(Single)</th>
        <th>FamilyType(Nuclear)</th>
        <th>FamilyType(Big)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>2</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>2</th>
        <td>2</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>3</th>
        <td>1</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>4</th>
        <td>2</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>5</th>
        <td>1</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>6</th>
        <td>1</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>7</th>
        <td>1</td>
        <td>True</td>
        <td>False</td>
        <td>False</td>
      </tr>
      <tr>
        <th>8</th>
        <td>5</td>
        <td>False</td>
        <td>False</td>
        <td>True</td>
      </tr>
      <tr>
        <th>9</th>
        <td>3</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
      <tr>
        <th>10</th>
        <td>2</td>
        <td>False</td>
        <td>True</td>
        <td>False</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 가족 수(FamilySize) 컬럼을 활용해 세 가지 구역을 나타내는 세 개의 새로운 컬럼을 만듭니다.
# 먼저 가족 수가 1명일 경우 Single 컬럼의 값에 True를, 1명이 아닐 경우 False를 대입합니다.
# 이후 가족 수가 2에서 4명 사이일 경우 Nuclear 컬럼의 값에 True를, 그렇지 않을 경우 False를 대입합니다.
# 마지막으로 가족 수가 5명 이상일 경우 Big 컬럼의 값에 True를, 그렇지 않을 경우 False를 대입합니다.
# FamilySize, FamilyType(Single), FamilyType(Nuclear), FamilyType(Big) 컬럼만 출력합니다.
print("### TYPE YOUR CODE ###")

In [79]:
train['FamilyType(Single)'] = train['FamilySize'] == 1
train['FamilyType(Nuclear)'] = (train['FamilySize'] >= 2) & (train['FamilySize'] < 5)
train['FamilyType(Big)'] = train['FamilySize'] >= 5

또는 위에서 생성한 FamilyType 컬럼을 사용하여 One Hot Encoding을 할 수 있습니다.

In [80]:
# 가족 형태(FamilyType)가 Single이면 True, 아니면 False인 리스트를 가져옵니다.
# FamilySize, FamilyType(Single), FamilyType(Nuclear), FamilyType(Big) 컬럼만 출력합니다.
print("### TYPE YOUR CODE ###")

### TYPE YOUR CODE ###


In [81]:
(train[['FamilySize', 'FamilyType(Single)', 'FamilyType(Nuclear)', 'FamilyType(Big)' ]]).head(10)

FamilySize  FamilyType(Single)  FamilyType(Nuclear)  \
PassengerId                                                        
1                     2               False                 True   
2                     2               False                 True   
3                     1                True                False   
4                     2               False                 True   
5                     1                True                False   
6                     1                True                False   
7                     1                True                False   
8                     5               False                False   
9                     3               False                 True   
10                    2               False                 True   

             FamilyType(Big)  
PassengerId                   
1                      False  
2                      False  
3                      False  
4                      False  
5                      False  
6                      False  
7                      False  
8                       True  
9                      False  
10                     False

<hr><span style="color:white">
train["FamilyType(Single)"] = train["FamilyType"] == "Single"
train["FamilyType(Nuclear)"] = train["FamilyType"] == "Nuclear"
train["FamilyType(Big)"] = train["FamilyType"] == "Big"
#print(train.shape)
train[["FamilySize", "FamilyType(Single)", "FamilyType(Nuclear)", "FamilyType(Big)"]].head(10)
</span><hr>

## 이름(Name) 컬럼 분석

**15. 이름(Name) 컬럼에서 호칭(Title) 컬럼을 뽑아주세요.**

이름(Name) 컬럼은 얼핏 분석하기 까다로워 보이지만, 실은 간단한 패턴으로 이루어져 있습니다. 일단 ```,```과 ```.```을 기준으로

  1. ```,```의 앞에 있는 부분이 성(SurName)이며,
  2. ```,```과 ```.``` 사이에 있는 부분은 승객의 호칭(Title)입니다.
  3. 마지막으로 ```.``` 뒤에 있는 부분은 이름(FirstName) 입니다.
  
가령 ```Braund, Mr. Owen Harris``` 이라는 이름의 승객은, 1) Braund라는 성을 가진, 2) Mr라는 호칭을 가진, 3) Owen Harris 라는 이름을 가진 승객으로 해석할 수 있습니다.

사용자의 이름(Name) 컬럼에서 **호칭(Title)** 컬럼을 추출해주세요. 최종적으로 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Name</th>
        <th>Title</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>Braund, Mr. Owen Harris</td>
        <td>Mr</td>
      </tr>
      <tr>
        <th>2</th>
        <td>Cumings, Mrs. John Bradley (Florence Briggs Th...</td>
        <td>Mrs</td>
      </tr>
      <tr>
        <th>3</th>
        <td>Heikkinen, Miss. Laina</td>
        <td>Miss</td>
      </tr>
      <tr>
        <th>4</th>
        <td>Futrelle, Mrs. Jacques Heath (Lily May Peel)</td>
        <td>Mrs</td>
      </tr>
      <tr>
        <th>5</th>
        <td>Allen, Mr. William Henry</td>
        <td>Mr</td>
      </tr>
      <tr>
        <th>6</th>
        <td>Moran, Mr. James</td>
        <td>Mr</td>
      </tr>
      <tr>
        <th>7</th>
        <td>McCarthy, Mr. Timothy J</td>
        <td>Mr</td>
      </tr>
      <tr>
        <th>8</th>
        <td>Palsson, Master. Gosta Leonard</td>
        <td>Master</td>
      </tr>
      <tr>
        <th>9</th>
        <td>Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)</td>
        <td>Mrs</td>
      </tr>
      <tr>
        <th>10</th>
        <td>Nasser, Mrs. Nicholas (Adele Achem)</td>
        <td>Mrs</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# get_title이라는 이름의 함수를 정의합니다. 이 함수는 name이라는 변수를 인자로 받습니다.
# 이 함수는 이름을 받았을 때 이름에서 타이틀을 반환해줍니다.
# 가령 name에 "Braund, Mr. Owen Harris"가 들어오면 최종 결과는 Mr를 반환해줍니다.

# 모든 Name 컬럼 데이터에 get_title 함수를 적용한 뒤 그 결과를 Title이라는 이름의 새로운 컬럼에 대입합니다.
# 이름(Name) 컬럼과 호칭(Title) 컬럼만을 출력합니다.
print("### TYPE YOUR CODE ###")

In [82]:
def get_title(name):
   
    title = name.split(',')[1].split('.')[0].strip()
    return title

In [83]:
train['Title'] = train['Name'].apply(get_title)

In [84]:
(train[['Name' ,'Title']]).head(10)

Name   Title
PassengerId                                                           
1                                      Braund, Mr. Owen Harris      Mr
2            Cumings, Mrs. John Bradley (Florence Briggs Th...     Mrs
3                                       Heikkinen, Miss. Laina    Miss
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)     Mrs
5                                     Allen, Mr. William Henry      Mr
6                                             Moran, Mr. James      Mr
7                                      McCarthy, Mr. Timothy J      Mr
8                               Palsson, Master. Gosta Leonard  Master
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)     Mrs
10                         Nasser, Mrs. Nicholas (Adele Achem)     Mrs

또한 호칭(Title) 별 생존률의 차이도 구해주세요.

In [ ]:
# pivot_table을 통해 호칭(Title)별 생존률을 출력합니다.
# Mr, Mrs, Master, Miss 등의 호칭별 생존률을 확인할 수 있습니다.
print("### TYPE YOUR CODE ###")

In [85]:
pivot_tavle6 = train.pivot_table(values= 'Survived', index= 'Title')
pivot_tavle6

Survived
Title                 
Capt          0.000000
Col           0.500000
Don           0.000000
Dr            0.428571
Jonkheer      0.000000
Lady          1.000000
Major         0.500000
Master        0.575000
Miss          0.697802
Mlle          1.000000
Mme           1.000000
Mr            0.156673
Mrs           0.792000
Ms            1.000000
Rev           0.000000
Sir           1.000000
the Countess  1.000000

<hr><span style="color:white">
pd.pivot_table(train, index="Title", values="Survived") 
</span><hr>

**16. 호칭(Title) 컬럼을 정리해주세요.**

호칭(Title) 컬럼은 생존자와 사망자를 구문하는데 중요한 역할을 담당합니다만, 너무 종류가 많은 단점이 있습니다. 

가령 호칭(Title) 컬럼은 Mr, Mrs, Miss, Master가 전체 승객의 호칭의 대부분을 차지하며, Capt, Col, Don, Sir 등과 같은 컬럼들은 모수가 작아서 오히려 신뢰할 수 없습니다.

가령 Sir라는 호칭을 가진 승객의 생존률이 100%라고 하더라도, 전체 승객애서 오직 1명만이 Sir라는 호칭을 사용하고 있기 때문에, "Sir라는 호칭을 사용하는 사람은 언제나 생존률이 높다."고 이야기하기에는 어렵습니다.



In [ ]:
# crosstab을 활용하여 각 호칭(Title)별 생존자의 총 인원수와 사망자의 총 인원수를 출력합니다.
print("### TYPE YOUR CODE ###")

In [86]:
cross_tab = pd.crosstab(train['Title'], train['Survived'])
cross_tab

Survived        0    1
Title                 
Capt            1    0
Col             1    1
Don             1    0
Dr              4    3
Jonkheer        1    0
Lady            0    1
Major           1    1
Master         17   23
Miss           55  127
Mlle            0    2
Mme             0    1
Mr            436   81
Mrs            26   99
Ms              0    1
Rev             6    0
Sir             0    1
the Countess    0    1

<hr><span style="color:white">
pd.crosstab(train["Title"], train["Survived"], margins=True)
</span><hr>

호칭(Title) 컬럼의 종류도 간소화 하려고 합니다. 호칭(Title) 컬럼에서 Mr, Mrs, Miss, Master를 제외한 나머지 값은 전부 **Other**라는 값으로 통일합니다. 최종적으로 하위 5개의 데이터에서는 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Name</th>
        <th>Title</th>
        <th>Title(Status)</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>887</th>
        <td>Montvila, Rev. Juozas</td>
        <td>Rev</td>
        <td>Others</td>
      </tr>
      <tr>
        <th>888</th>
        <td>Graham, Miss. Margaret Edith</td>
        <td>Miss</td>
        <td>Miss</td>
      </tr>
      <tr>
        <th>889</th>
        <td>Johnston, Miss. Catherine Helen "Carrie"</td>
        <td>Miss</td>
        <td>Miss</td>
      </tr>
      <tr>
        <th>890</th>
        <td>Behr, Mr. Karl Howell</td>
        <td>Mr</td>
        <td>Mr</td>
      </tr>
      <tr>
        <th>891</th>
        <td>Dooley, Mr. Patrick</td>
        <td>Mr</td>
        <td>Mr</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 이번에는 get_cleaned_title 이름의 함수를 정의합니다.
# 이 함수에서는 Mr, Mrs, Miss, Master 가 아닌 호칭은 전부 Other라고 반환합니다.

# 모든 Name 컬럼 데이터에 get_cleaned_title 함수를 적용한 뒤
# 그 결과를 Title(Clean)이라는 이름의 새로운 컬럼에 대입합니다.
# 이름(Name) 컬럼과 Title 컬럼, 그리고 Title(Clean) 컬럼만을 출력합니다.
# 다만 이번에는 head가 아니라 tail을 활용해 마지막 다섯 컬럼만 출력합니다.
print("### TYPE YOUR CODE ###")

In [88]:
def get_cleaned_title(title):
    if title in ['Mr', 'Mrs', 'Miss', 'Master']:
        return title
    else:
        return 'Others'

In [89]:
train['Title(Status)'] = train['Title'].apply(get_cleaned_title)

In [90]:
(train[['Name' ,'Title', 'Title(Status)']]).tail()

Name Title Title(Status)
PassengerId                                                              
887                             Montvila, Rev. Juozas   Rev        Others
888                      Graham, Miss. Margaret Edith  Miss          Miss
889          Johnston, Miss. Catherine Helen "Carrie"  Miss          Miss
890                             Behr, Mr. Karl Howell    Mr            Mr
891                               Dooley, Mr. Patrick    Mr            Mr

In [ ]:
### strip 테스트
#train["Title"][1].strip()

또한 마찬가지로 새롭게 정리한 호칭별 생존률의 차이도 구해주세요.

In [ ]:
# pivot_table을 통해 정리한 호칭별 생존률을 출력합니다.
# Mr, Mrs, Master, Miss, Others별 생존률을 확인할 수 있습니다.
print("### TYPE YOUR CODE ###")

In [91]:
pivot_tavle7 = train.pivot_table(values= 'Survived', index= 'Title(Status)')
pivot_tavle7

Survived
Title(Status)          
Master         0.575000
Miss           0.697802
Mr             0.156673
Mrs            0.792000
Others         0.444444

In [92]:
pivot_tavle7 * 100

Survived
Title(Status)           
Master         57.500000
Miss           69.780220
Mr             15.667311
Mrs            79.200000
Others         44.444444

<hr><span style="color:white">
pd.pivot_table(train, index="Title(Status)", values="Survived")
</span><hr>

**17. 호칭(Title) 컬럼으로 결혼 유무를 파악해주세요.**

앞서 우리가 추출한 호칭 컬럼의 주요 값에 대한 설명은 다음과 같습니다.

  * Mr - 결혼한 남성을 의미합니다.
  * Master - 결혼하지 않은 남성을 의미합니다.
  * Mrs - 결혼한 여성을 의미합니다.
  * Miss - 결혼하지 않은 여성을 의미합니다.
  
그러므로 호칭(Title) 컬럼값에서는 추가로 결혼 여부를 뽑아낼 수 있습니다. 호칭이 Mr이거나 Mrs이면 결혼했다고 간주하고, Master이거나 Miss이면 결혼하지 않았다고 간주할 수 있습니다. (나머지 호칭은 NaN을 집어넣으면 좋겠습니다)

호칭(Title) 컬럼을 활용하여 **결혼 유무(Married)**를 추출해주세요, 최종적으로 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Name</th>
        <th>Title</th>
        <th>Married</th>
      </tr>
      <tr>
        <th>PassengerId</th>
        <th></th>
        <th></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>1</th>
        <td>Braund, Mr. Owen Harris</td>
        <td>Mr</td>
        <td>True</td>
      </tr>
      <tr>
        <th>2</th>
        <td>Cumings, Mrs. John Bradley (Florence Briggs Th...</td>
        <td>Mrs</td>
        <td>True</td>
      </tr>
      <tr>
        <th>3</th>
        <td>Heikkinen, Miss. Laina</td>
        <td>Miss</td>
        <td>False</td>
      </tr>
      <tr>
        <th>4</th>
        <td>Futrelle, Mrs. Jacques Heath (Lily May Peel)</td>
        <td>Mrs</td>
        <td>True</td>
      </tr>
      <tr>
        <th>5</th>
        <td>Allen, Mr. William Henry</td>
        <td>Mr</td>
        <td>True</td>
      </tr>
      <tr>
        <th>6</th>
        <td>Moran, Mr. James</td>
        <td>Mr</td>
        <td>True</td>
      </tr>
      <tr>
        <th>7</th>
        <td>McCarthy, Mr. Timothy J</td>
        <td>Mr</td>
        <td>True</td>
      </tr>
      <tr>
        <th>8</th>
        <td>Palsson, Master. Gosta Leonard</td>
        <td>Master</td>
        <td>False</td>
      </tr>
      <tr>
        <th>9</th>
        <td>Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)</td>
        <td>Mrs</td>
        <td>True</td>
      </tr>
      <tr>
        <th>10</th>
        <td>Nasser, Mrs. Nicholas (Adele Achem)</td>
        <td>Mrs</td>
        <td>True</td>
      </tr>
    </tbody>
  </table>

또한 결혼 유무(Married)별 생존률의 차이도 구해주세요. 최종적으로 다음의 결과가 나와야 합니다.

<table border="1" class="dataframe" style="float: left;">
    <thead>
      <tr style="text-align: right;">
        <th></th>
        <th>Survived</th>
      </tr>
      <tr>
        <th>Married</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>False</th>
        <td>0.675676</td>
      </tr>
      <tr>
        <th>True</th>
        <td>0.280374</td>
      </tr>
    </tbody>
  </table>

In [ ]:
# 4개의 Title 컬럼을 활용한 결혼 유무에 따른 Married 컬럼 생성
print("### TYPE YOUR CODE ###")

In [93]:
train["Title"] = train["Title"].str.strip()
train.loc[train["Title"] == "Mr", "Married"] = True
train.loc[train["Title"] == "Master", "Married"] = False
train.loc[train["Title"] == "Mrs", "Married"] = True
train.loc[train["Title"] == "Miss", "Married"] = False

In [ ]:
# 이름(Name) 컬럼과 호칭(Title) 컬럼, 그리고 결혼 유무(Married) 컬럼을 출력합니다.
print("### TYPE YOUR CODE ###")

In [97]:
train[["Name", "Title", "Married"]].head(10)

Name   Title Married
PassengerId                                                                   
1                                      Braund, Mr. Owen Harris      Mr    True
2            Cumings, Mrs. John Bradley (Florence Briggs Th...     Mrs    True
3                                       Heikkinen, Miss. Laina    Miss   False
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)     Mrs    True
5                                     Allen, Mr. William Henry      Mr    True
6                                             Moran, Mr. James      Mr    True
7                                      McCarthy, Mr. Timothy J      Mr    True
8                               Palsson, Master. Gosta Leonard  Master   False
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)     Mrs    True
10                         Nasser, Mrs. Nicholas (Adele Achem)     Mrs    True

또는 isin() 함수를 사용하여 결혼 유무를 파악할 수 있습니다.

In [ ]:
# 호칭(Title)이 Mr이거나 Mrs인 데이터를 가져옵니다.
# 이후 결혼 유무(Married)라는 새로운 컬럼에 True라는 값을 대입합니다.

# 비슷하게 호칭(Title)이 Master이거나 Miss인 데이터를 가져옵니다.
# 마찬가지로 결혼 유무(Married)라는 컬럼에 False라는 값을 대입합니다.
# 이름(Name) 컬럼과 호칭(Title) 컬럼, 그리고 결혼 유무(Married) 컬럼을 출력합니다.
print("### TYPE YOUR CODE ###")

In [95]:
train.loc[train["Title(Status)"].isin(["Mr", "Mrs"]), "Married"] = True
train.loc[train["Title(Status)"].isin(["Master", "Miss"]), "Married"] = False
train.loc[train["Title(Status)"].isin(["Others"]), "Married"] = False

In [96]:
train[["Name", "Title", "Married"]].head(10)

Name   Title Married
PassengerId                                                                   
1                                      Braund, Mr. Owen Harris      Mr    True
2            Cumings, Mrs. John Bradley (Florence Briggs Th...     Mrs    True
3                                       Heikkinen, Miss. Laina    Miss   False
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)     Mrs    True
5                                     Allen, Mr. William Henry      Mr    True
6                                             Moran, Mr. James      Mr    True
7                                      McCarthy, Mr. Timothy J      Mr    True
8                               Palsson, Master. Gosta Leonard  Master   False
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)     Mrs    True
10                         Nasser, Mrs. Nicholas (Adele Achem)     Mrs    True

In [ ]:
# pivot_table을 통해 결혼 유무(Married)별 따른 생존률을 출력합니다.
# 결혼을 한 승객(True)와 하지 않은 승객(False)간의 생존률의 차이를 볼 수 있습니다.
print("### TYPE YOUR CODE ###")

In [98]:
pd.pivot_table(train, index="Married", values="Survived")

Survived
Married          
False    0.650602
True     0.280374

<hr><span style="color:white">
pd.pivot_table(train, index="Married", values="Survived")
</span><hr>

In [ ]:
train.to_csv('data_titanic/train_one_hot.csv')

## 마무리하며

pandas 학습에 참고할만한 추천 자료:

  * [10 minutes to pandas](https://pandas.pydata.org/pandas-docs/stable/10min.html)
  * [Pandas Cookbook](http://github.com/jvns/pandas-cookbook)
  * [Python for Data Science](http://wavedatalab.github.io/datawithpython/)
  * [Modern Pandas](http://tomaugspurger.github.io/modern-1-intro.html)
